client/src/tpu_client.rs

```diff
@@ -0,0 +1,393 @@
+        // so check if the current slot is in line with the recent progression.
+        let max_index = recent_slots.len() - 1;
+        let median_index = max_index / 2;
+        let median_recent_slot = recent_slots[median_index];
+        let expected_current_slot = median_recent_slot + (max_index - median_index) as u64;
+        let max_reasonable_current_slot = expected_current_slot + MAX_SLOT_SKIP_DISTANCE;
+
+        // Return the highest slot that doesn't exceed what we believe is a
+        // reasonable slot.
+        recent_slots
+            .into_iter()
+            .rev()
+            .find(|slot| *slot <= max_reasonable_current_slot)
+            .unwrap()
+    }
+}
+
+#[cfg(test)]
+impl From<Vec<Slot>> for RecentLeaderSlots {
+    fn from(recent_slots: Vec<Slot>) -> Self {
+        assert!(!recent_slots.is_empty());
+        Self(Arc::new(RwLock::new(recent_slots.into_iter().collect())))
+    }
+}
+
+/// Service that tracks upcoming leaders and maintains an up-to-date mapping
+/// of leader id to TPU socket address.
+struct LeaderTpuService {
+    recent_slots: RecentLeaderSlots,
+    leader_tpu_cache: Arc<RwLock<LeaderTpuCache>>,
+    subscription: Option<PubsubClientSubscription<SlotUpdate>>,
+    t_leader_tpu_service: Option<JoinHandle<()>>,
+}
+
+impl LeaderTpuService {
+    fn new(rpc_client: Arc<RpcClient>, websocket_url: &str, exit: Arc<AtomicBool>) -> Result<Self> {
+        let start_slot = rpc_client.get_max_shred_insert_slot()?;
+
+        let recent_slots = RecentLeaderSlots::new(start_slot);
+        let leader_tpu_cache = Arc::new(RwLock::new(LeaderTpuCache::new(&rpc_client, start_slot)));
+
+        let subscription = if !websocket_url.is_empty() {
+            let recent_slots = recent_slots.clone();
+            Some(PubsubClient::slot_updates_subscribe(
+                websocket_url,
+                move |update| {
+                    let current_slot = match update {
+                        // This update indicates that a full slot was received by the connected
+                        // node so we can stop sending transactions to the leader for that slot
+                        SlotUpdate::Completed { slot, .. } => slot.saturating_add(1),
+                        // This update indicates that we have just received the first shred from
+                        // the leader for this slot and they are probably still accepting transactions.
+                        SlotUpdate::FirstShredReceived { slot, .. } => slot,
+                        _ => return,
+                    };
+
+                    recent_slots.record_slot(current_slot);
+                },
+            )?)
+        } else {
+            None
+        };
+
+        let t_leader_tpu_service = Some({
+            let recent_slots = recent_slots.clone();
+            let leader_tpu_cache = leader_tpu_cache.clone();
+            std::thread::Builder::new()
+                .name("solana-leader-tpu-service".to_string())
```

Comment:
The cryptic `ldr-tpu-srv` is ironically probably a more descriptive linux thread name than `solana-leader-`. I believe we get 15 characters